src/testing_tracer.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    fn reset_exception(&mut self) {
+        self.exception_message = None;
+    }
+
+    fn set_exception_message(&mut self, message: &str) {
+        self.exception_message = Some(message.to_owned());
+    }
+
+    fn execute_print(&mut self, message: &str) {
+        println!("{}", message);
+    }
+
+    fn execute_print_from_register(&mut self, val: PrimitiveValue) {
+        assert!(
+            self.expecting_register_value,
+            "Unexpected print_from_register command"
+        );
+        self.expecting_register_value = false;
+        println!("{}", val.value);
+    }
+
+    fn parse_command_from_register(
+        &mut self,
+        val: PrimitiveValue,
+    ) -> (Option<String>, Option<String>) {
+        if val.value == U256::from(0) {
+            return (None, None);
+        }
+
+        let bytes: &mut [u8; 32] = &mut [0; 32];
+        val.value.to_big_endian(bytes);
+
+        if let Ok(message) = std::str::from_utf8(bytes) {
+            let message_trimmed = message.trim_matches(char::from(0));
+
+            for prefix in [
+                EXCEPTION_PREFIX,
+                PRINT_PREFIX,
+                PRINT_REG_PREFIX,
+                PRINT_PTR_PREFIX,
+            ]
+            .iter()
+            {
+                if message_trimmed.starts_with(*prefix) {
+                    let arg = message_trimmed.strip_prefix(prefix).unwrap();
+                    return (Some((*prefix).to_string()), Some(arg.to_owned()));
+                }
+            }
+        }
+
+        return (None, None);
+    }
+}
+
+impl Tracer for TestingTracer {
+    type SupportedMemory = SimpleMemory;
+    const CALL_BEFORE_EXECUTION: bool = true;
+    const CALL_AFTER_DECODING: bool = true;
+
+    #[inline]
+    fn before_decoding(&mut self, _state: VmLocalStateData<'_>, _memory: &Self::SupportedMemory) {}
+
+    fn after_decoding(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: AfterDecodingData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        // check for built-in panics
+        if !_data.error_flags_accumulated.is_empty() {
+            // last accumulated panic will be used as exception_message
+            for (panic, _) in _data.error_flags_accumulated.iter_names() {
+                self.exception_message = Some(panic.to_owned());
+            }
+        }
+    }
+
+    fn before_execution(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: BeforeExecutionData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        let inner_opcode = _data.opcode.inner.variant.opcode;
+
+        // Propagate error message if Nop, ret.panic, ret.revert; reset otherwise
+        match inner_opcode {
+            Opcode::Nop(NopOpcode) => {}
+            Opcode::Ret(RetOpcode::Panic) => {}
+            Opcode::Ret(RetOpcode::Revert) => {}
+            _ => {
+                self.reset_exception();
+            }
+        }
+
+        // Try to execute commands
```

Comment:
nit: I'd improve this comment - to say that you're trying to see, if a special command is present.